src/ast/irbuilderbpf.cpp

```diff
@@ -698,46 +712,108 @@ Value *IRBuilderBPF::CreatePerCpuMapAggElems(Value *ctx,
               ret);
 }
 
 void IRBuilderBPF::createPerCpuMinMax(AllocaInst *ret,
-                                      Value *cpu_value,
-                                      bool is_max)
+                                      AllocaInst *is_ret_set,
+                                      CallInst *call,
+                                      const SizedType &type)
 {
+  auto *value_type = GetMapValueType(type);
+  auto *cast = CreatePointerCast(call, value_type->getPointerTo(), "cast");
+  bool is_max = type.IsMaxTy();
+
+  Value *mm_val = CreateLoad(
+      getInt64Ty(), CreateGEP(value_type, cast, { getInt64(0), getInt32(0) }));
+
+  Value *is_val_set = CreateLoad(
+      getInt64Ty(), CreateGEP(value_type, cast, { getInt64(0), getInt32(1) }));
+
+  /*
+   * (ret, is_ret_set, min_max_val, is_val_set) {
+   * // if the min_max_val is 0, which is the initial map value,
+   * // we need to know if it was explicitly set by user
+   * if (!is_val_set == 1) {
+   *   return;
+   * }
+   * if (!is_ret_set == 1) {
+   *   ret = min_max_val;
+   *   is_ret_set = 1;
+   * } else if (min_max_val > ret) { // or min_max_val < ret if min operation
+   *   ret = min_max_val;
+   *   is_ret_set = 1;
+   * }
+   */
+
   Function *parent = GetInsertBlock()->getParent();
-  BasicBlock *success_block = BasicBlock::Create(module_.getContext(),
-                                                 "min_max_success",
-                                                 parent);
+  BasicBlock *val_set_success = BasicBlock::Create(module_.getContext(),
+                                                   "val_set_success",
+                                                   parent);
+  BasicBlock *min_max_success = BasicBlock::Create(module_.getContext(),
+                                                   "min_max_success",
+                                                   parent);
+  BasicBlock *ret_set_success = BasicBlock::Create(module_.getContext(),
+                                                   "ret_set_success",
+                                                   parent);
   BasicBlock *merge_block = BasicBlock::Create(module_.getContext(),
                                                "min_max_merge",
                                                parent);
-  Value *condition;
+
+  Value *val_set_condition = CreateICmpEQ(is_val_set,
+                                          getInt64(1),
+                                          "val_set_cond");
+
+  Value *ret_set_condition = CreateICmpEQ(CreateLoad(getInt64Ty(), is_ret_set),
+                                          getInt64(1),
+                                          "ret_set_cond");
+
+  Value *min_max_condition;
 
   if (is_max) {
-    condition = CreateICmpSGT(CreateLoad(getInt64Ty(), cpu_value),
-                              CreateLoad(getInt64Ty(), ret),
-                              "max_cond");
+    min_max_condition = CreateICmpSGT(mm_val,
```

Comment:
Do these min_min_conditions need to be made to handle unsigned values too?

Reply:
Ah good catch. I'll update.